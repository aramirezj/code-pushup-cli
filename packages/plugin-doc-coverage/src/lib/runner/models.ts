import type { SyntaxKind } from 'ts-morph';

/** Maps the SyntaxKind from the library ts-morph to the coverage type. */
type SyntaxKindToStringLiteral = {
  [SyntaxKind.ClassDeclaration]: 'classes';
  [SyntaxKind.MethodDeclaration]: 'methods';
  [SyntaxKind.FunctionDeclaration]: 'functions';
  [SyntaxKind.InterfaceDeclaration]: 'interfaces';
  [SyntaxKind.EnumDeclaration]: 'enums';
  [SyntaxKind.VariableDeclaration]: 'variables';
  [SyntaxKind.PropertyDeclaration]: 'properties';
  [SyntaxKind.TypeAliasDeclaration]: 'types';
};

/**The coverage type is the same as the SyntaxKind from the library ts-morph but as a string. */
export type CoverageType =
  SyntaxKindToStringLiteral[keyof SyntaxKindToStringLiteral];

/** The undocumented node is the node that is not documented and has the information for the report. */
export type UndocumentedNode = {
  file: string;
  type: CoverageType;
  name: string;
  line: number;
  class?: string;
};

/** The coverage data is the data that is used to create the coverage report. Without coverage stats yet */
export type CoverageData = {
  issues: UndocumentedNode[];
  nodesCount: number;
};

/** The unprocessed coverage result CoverageData but for each coverage type. */
export type UnprocessedCoverageResult = Record<CoverageType, CoverageData>;

/** The processed coverage result CoverageData but for each coverage type and with coverage stats. */
export type CoverageResult = Record<
  CoverageType,
  CoverageData & {
    coverage: number;
  }
>;
