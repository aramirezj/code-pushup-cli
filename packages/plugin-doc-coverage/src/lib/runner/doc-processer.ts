import {
  ClassDeclaration,
  Project,
  SourceFile,
  VariableStatement,
} from 'ts-morph';
import type { DocCoveragePluginConfig } from '../config.js';
import type {
  CoverageResult,
  CoverageType,
  UnprocessedCoverageResult,
} from './models.js';
import {
  calculateCoverage,
  createEmptyUnprocessedCoverageReport,
  getCoverageTypeFromKind,
} from './utils.js';

/**
 * Gets the variables information from the variable statements
 * @param variableStatements - The variable statements to process
 * @returns {Node[]} The variables information with the right methods to get the information
 */
export function getVariablesInformation(
  variableStatements: VariableStatement[],
) {
  return variableStatements.flatMap(variable => {
    // Get parent-level information
    const parentInfo = {
      getKind: () => variable.getKind(),
      getJsDocs: () => variable.getJsDocs(),
      getStartLineNumber: () => variable.getStartLineNumber(),
    };

    // Map each declaration to combine parent info with declaration-specific info
    return variable.getDeclarations().map(declaration => ({
      ...parentInfo,
      getName: () => declaration.getName(),
    }));
  });
}

/**
 * Processes documentation coverage for TypeScript files in the specified path
 * @param toInclude - The file path pattern to include for documentation analysis
 * @returns {CoverageResult} Object containing coverage statistics and undocumented items
 */
export function processDocCoverage(
  config: DocCoveragePluginConfig,
): CoverageResult {
  const project = new Project();
  project.addSourceFilesAtPaths(config.sourceGlob);
  return getUnprocessedCoverageReport(project.getSourceFiles());
}

/**
 * Gets the unprocessed coverage report from the source files
 * @param sourceFiles - The source files to process
 * @returns {UnprocessedCoverageResult} The unprocessed coverage report
 */
export function getUnprocessedCoverageReport(
  sourceFiles: SourceFile[],
): CoverageResult {
  const unprocessedCoverageReport = sourceFiles.reduce(
    (coverageReportOfAllFiles, sourceFile) => {
      const filePath = sourceFile.getFilePath();
      const classes = sourceFile.getClasses();

      const allNodesFromFile = [
        ...sourceFile.getFunctions(),
        ...classes,
        ...getClassNodes(classes),
        ...sourceFile.getTypeAliases(),
        ...sourceFile.getEnums(),
        ...sourceFile.getInterfaces(),
        ...getVariablesInformation(sourceFile.getVariableStatements()),
      ];

      const coverageReportOfCurrentFile = allNodesFromFile.reduce(
        (acc, node) => {
          const nodeType = getCoverageTypeFromKind(node.getKind());
          const currentTypeReport = acc[nodeType];

          const updatedIssues =
            node.getJsDocs().length === 0
              ? [
                  ...currentTypeReport.issues,
                  {
                    file: filePath,
                    type: nodeType,
                    name: node.getName() || '',
                    line: node.getStartLineNumber(),
                  },
                ]
              : currentTypeReport.issues;

          return {
            ...acc,
            [nodeType]: {
              nodesCount: currentTypeReport.nodesCount + 1,
              issues: updatedIssues,
            },
          };
        },
        createEmptyUnprocessedCoverageReport(),
      );

      return mergeCoverageResults(
        coverageReportOfAllFiles,
        coverageReportOfCurrentFile,
      );
    },
    createEmptyUnprocessedCoverageReport(),
  );

  return calculateCoverage(unprocessedCoverageReport);
}

/**
 * Merges two coverage results
 * @param results - The first empty coverage result
 * @param current - The second coverage result
 * @returns {UnprocessedCoverageResult} The merged coverage result
 */
export function mergeCoverageResults(
  results: UnprocessedCoverageResult,
  current: Partial<UnprocessedCoverageResult>,
): UnprocessedCoverageResult {
  return Object.fromEntries(
    Object.entries(results).map(([key, value]) => {
      const node = value as CoverageResult[CoverageType];
      const type = key as CoverageType;
      return [
        type,
        {
          nodesCount: node.nodesCount + (current[type]?.nodesCount ?? 0),
          issues: [...node.issues, ...(current[type]?.issues ?? [])],
        },
      ];
    }),
  ) as UnprocessedCoverageResult;
}

/**
 * Gets the nodes from a class
 * @param classNodes - The class nodes to process
 * @returns {Node[]} The nodes from the class
 */
export function getClassNodes(classNodes: ClassDeclaration[]) {
  return classNodes.flatMap(classNode => [
    ...classNode.getMethods(),
    ...classNode.getProperties(),
  ]);
}
