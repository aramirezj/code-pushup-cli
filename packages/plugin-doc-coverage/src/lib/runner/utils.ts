import { SyntaxKind } from 'ts-morph';
import type {
  CoverageReportShape,
  CoverageResult,
  CoverageType,
} from './models.js';

/**
 * Creates an empty unprocessed coverage report.
 * @returns The empty unprocessed coverage report.
 */
export function createEmptyCoverageData(): CoverageReportShape {
  return {
    enums: { nodesCount: 0, issues: [] },
    interfaces: { nodesCount: 0, issues: [] },
    types: { nodesCount: 0, issues: [] },
    functions: { nodesCount: 0, issues: [] },
    variables: { nodesCount: 0, issues: [] },
    classes: { nodesCount: 0, issues: [] },
    methods: { nodesCount: 0, issues: [] },
    properties: { nodesCount: 0, issues: [] },
  };
}

/**
 * Calculates the coverage percentage for each coverage type.
 * @param result - The unprocessed coverage result.
 * @returns The processed coverage result.
 */
export function calculateCoverage(result: CoverageReportShape) {
  return Object.fromEntries(
    Object.entries(result).map(([key, value]) => {
      const type = key as CoverageType;
      return [
        type,
        {
          coverage:
            value.nodesCount === 0
              ? 100
              : Number(
                  ((1 - value.issues.length / value.nodesCount) * 100).toFixed(
                    2,
                  ),
                ),
          issues: value.issues,
          nodesCount: value.nodesCount,
        },
      ];
    }),
  ) as CoverageResult;
}

/**
 * Maps the SyntaxKind from the library ts-morph to the coverage type.
 * @param kind - The SyntaxKind from the library ts-morph.
 * @returns The coverage type.
 */
export function getCoverageTypeFromKind(kind: SyntaxKind): CoverageType {
  switch (kind) {
    case SyntaxKind.ClassDeclaration:
      return 'classes';
    case SyntaxKind.MethodDeclaration:
      return 'methods';
    case SyntaxKind.FunctionDeclaration:
      return 'functions';
    case SyntaxKind.InterfaceDeclaration:
      return 'interfaces';
    case SyntaxKind.EnumDeclaration:
      return 'enums';
    case SyntaxKind.VariableStatement:
    case SyntaxKind.VariableDeclaration:
      return 'variables';
    case SyntaxKind.PropertyDeclaration:
      return 'properties';
    case SyntaxKind.TypeAliasDeclaration:
      return 'types';
    default:
      throw new Error(`Unsupported syntax kind: ${kind}`);
  }
}
