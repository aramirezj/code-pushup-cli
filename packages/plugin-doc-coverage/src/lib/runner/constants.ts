import path from 'node:path';
import { pluginWorkDir } from '@code-pushup/utils';

export const WORKDIR = pluginWorkDir('doc-coverage');

export const PLUGIN_CONFIG_PATH = path.join(
  process.cwd(),
  WORKDIR,
  'plugin-config.json',
);
export const RUNNER_OUTPUT_PATH = path.join(WORKDIR, 'runner-output.json');
