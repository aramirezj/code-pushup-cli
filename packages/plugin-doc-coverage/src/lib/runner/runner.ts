import type { AuditOutputs, RunnerFunction } from '@code-pushup/models';
import type { DocCoveragePluginConfig } from '../config.js';
import { processDocCoverage } from './doc-processer.js';
import type { CoverageType, DocumentationCoverageReport } from './models.js';
import { coverageTypeToAuditSlug } from './utils.js';

export function createRunnerFunction(
  config: DocCoveragePluginConfig,
): RunnerFunction {
  return (): AuditOutputs => {
    const coverageResult = processDocCoverage(config);
    return trasformCoverageReportToAuditOutputs(coverageResult, config);
  };
}

/**
 * Transforms the coverage report into audit outputs.
 * @param coverageResult - The coverage result containing undocumented items and coverage statistics
 * @param options - Configuration options specifying which audits to include and exclude
 * @returns Audit outputs with coverage scores and details about undocumented items
 */
export function trasformCoverageReportToAuditOutputs(
  coverageResult: DocumentationCoverageReport,
  options: Pick<DocCoveragePluginConfig, 'onlyAudits' | 'skipAudits'>,
): AuditOutputs {
  return Object.entries(coverageResult)
    .filter(([type]) => {
      const auditSlug = coverageTypeToAuditSlug(type as CoverageType);
      if (options.onlyAudits?.length) {
        return options.onlyAudits.includes(auditSlug);
      }
      if (options.skipAudits?.length) {
        return !options.skipAudits.includes(auditSlug);
      }
      return true;
    })
    .map(([type, item]) => {
      const { coverage, issues } = item;

      return {
        slug: `${type}-coverage`,
        value: issues.length,
        score: coverage / 100,
        displayValue: `${issues.length} undocumented ${type}`,
        details: {
          issues: item.issues.map(({ file, line, name }) => ({
            message: `Missing ${type} documentation for ${name}`,
            source: { file, position: { startLine: line } },
            severity: 'warning',
          })),
        },
      };
    });
}
