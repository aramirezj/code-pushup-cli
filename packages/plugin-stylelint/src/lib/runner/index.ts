import { LinterOptions } from 'stylelint';
import type { RunnerFunction } from '@code-pushup/models';
import { lintStyles } from './stylelint-runner.js';
import { mapStylelintResultsToAudits } from './utils';

export function createRunnerFunction(opt: LinterOptions): RunnerFunction {
  return async () => {
    const report = await lintStyles(opt);
    return mapStylelintResultsToAudits(report);
  };
}
