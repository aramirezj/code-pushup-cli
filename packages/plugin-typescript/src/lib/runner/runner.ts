import type {
  Audit,
  AuditOutput,
  AuditOutputs,
  AuditReport,
  Issue,
  RunnerFunction,
} from '@code-pushup/models';
import {
  type DiagnosticsOptions,
  getTypeScriptDiagnostics,
} from './ts-runner.js';
import type { CodeRangeName } from './types.js';
import { getIssueFromDiagnostic, tSCodeToAuditSlug } from './utils.js';

export type RunnerOptions = DiagnosticsOptions & {
  expectedAudits: Pick<Audit, 'slug'>[];
};

export function createRunnerFunction(options: RunnerOptions): RunnerFunction {
  const { tsConfigPath, expectedAudits } = options;
  return async (): Promise<AuditOutputs> => {
    const diagnostics = await getTypeScriptDiagnostics({ tsConfigPath });
    const result: Record<
      CodeRangeName,
      Pick<AuditReport, 'slug' | 'details'>
    > = diagnostics.reduce(
      (acc, diag) => {
        const slug = tSCodeToAuditSlug(diag.code);
        const existingIssues: Issue[] =
          (acc[slug] && acc[slug].details?.issues) || ([] as Issue[]);
        return {
          ...acc,
          [slug]: {
            slug,
            details: {
              issues: [...existingIssues, getIssueFromDiagnostic(diag)],
            },
          },
        };
      },
      {} as unknown as Record<
        CodeRangeName,
        Pick<AuditOutput, 'slug' | 'details'>
      >,
    );

    return expectedAudits.map(({ slug }) => {
      const { details } = result[slug as CodeRangeName] ?? {};

      const issues = details?.issues ?? [];
      return {
        slug,
        score: issues.length === 0 ? 1 : 0,
        value: issues.length,
        ...(issues.length > 0 ? { details } : {}),
      } satisfies AuditOutput;
    });
  };
}
