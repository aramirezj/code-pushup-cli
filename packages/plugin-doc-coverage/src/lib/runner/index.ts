import { bold } from 'ansis';
import { writeFile } from 'node:fs/promises';
import path from 'node:path';
import type { AuditOutput, RunnerConfig } from '@code-pushup/models';
import {
  ProcessError,
  ensureDirectoryExists,
  executeProcess,
  filePathToCliArg,
  readJsonFile,
  ui,
} from '@code-pushup/utils';
import type { DocCoveragePluginConfig } from '../config.js';
import { PLUGIN_CONFIG_PATH, RUNNER_OUTPUT_PATH } from './constants.js';

export { PLUGIN_CONFIG_PATH, RUNNER_OUTPUT_PATH } from './constants.js';

export async function executeRunner(): Promise<void> {
  const { outputPath, coverageToolCommand } =
    await readJsonFile<DocCoveragePluginConfig>(PLUGIN_CONFIG_PATH);
  if (coverageToolCommand != null) {
    const { command, args = [] } = coverageToolCommand;
    try {
      await executeProcess({ command, args });
    } catch (error) {
      if (error instanceof ProcessError) {
        ui().logger.error(bold('stdout from failed Compodoc process:'));
        ui().logger.error(error.stdout);
        ui().logger.error(bold('stderr from failed Compodoc process:'));
        ui().logger.error(error.stderr);
      }
      throw new Error(
        'Doc Coverage plugin: Running Compodoc failed. Please check the error above.',
      );
    }
  }

  try {
    // From the output of Compodoc, we can get the coverage percentage.
    const docData: { coverage: { count: number } } =
      await readJsonFile(outputPath);
    const coverage = docData.coverage.count || 0;

    const auditOutputs: AuditOutput[] = [
      {
        slug: 'percentage-coverage',
        value: coverage,
        score: coverage / 100,
        displayValue: `${coverage} %`,
      },
    ];

    await ensureDirectoryExists(path.dirname(RUNNER_OUTPUT_PATH));
    await writeFile(RUNNER_OUTPUT_PATH, JSON.stringify(auditOutputs));
  } catch (error) {
    if (error instanceof ProcessError) {
      ui().logger.error(bold('stdout from failed coverage tool process:'));
      ui().logger.error(error.stdout);
      ui().logger.error(bold('stderr from failed coverage tool process:'));
      ui().logger.error(error.stderr);

      throw new Error(
        'Doc Coverage plugin: Running Compodoc failed. Please check the error above.',
      );
    }
  }
}

export async function createRunnerConfig(
  scriptPath: string,
  config: DocCoveragePluginConfig,
): Promise<RunnerConfig> {
  // Create JSON config for executeRunner
  await ensureDirectoryExists(path.dirname(PLUGIN_CONFIG_PATH));
  await writeFile(PLUGIN_CONFIG_PATH, JSON.stringify(config));

  return {
    command: 'node',
    args: [filePathToCliArg(scriptPath)],
    outputFile: RUNNER_OUTPUT_PATH,
  };
}
