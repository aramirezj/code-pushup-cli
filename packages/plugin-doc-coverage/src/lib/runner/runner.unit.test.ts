import type { CoverageResult } from './models.js';
import { trasformCoverageReportToAudits } from './runner.js';

describe('trasformCoverageReportToAudits', () => {
  const mockCoverageResult = {
    functions: {
      coverage: 75,
      nodesCount: 4,
      issues: [
        {
          file: 'test.ts',
          line: 10,
          name: 'testFunction',
          type: 'functions',
        },
      ],
    },
    classes: {
      coverage: 100,
      nodesCount: 2,
      issues: [],
    },
  } as unknown as CoverageResult;

  it('should transform coverage report to audit outputs with no filters', () => {
    const result = trasformCoverageReportToAudits(mockCoverageResult, {});
    expect(result).toMatchSnapshot();
  });

  it('should filter audits when onlyAudits is provided', () => {
    const result = trasformCoverageReportToAudits(mockCoverageResult, {
      onlyAudits: ['functions-coverage'],
    });
    expect(result).toMatchSnapshot();
  });

  it('should filter audits when skipAudits is provided', () => {
    const result = trasformCoverageReportToAudits(mockCoverageResult, {
      skipAudits: ['classes-coverage'],
    });
    expect(result).toMatchSnapshot();
  });

  it('should handle empty coverage result', () => {
    const result = trasformCoverageReportToAudits(
      {} as unknown as CoverageResult,
      {},
    );
    expect(result).toMatchSnapshot();
  });

  it('should handle coverage result with multiple issues', () => {
    const coverageWithMultipleIssues = {
      functions: {
        coverage: 50,
        nodesCount: 4,
        issues: [
          {
            file: 'test1.ts',
            line: 10,
            name: 'function1',
            type: 'functions',
          },
          {
            file: 'test2.ts',
            line: 20,
            name: 'function2',
            type: 'functions',
          },
        ],
      },
    } as unknown as CoverageResult;

    const result = trasformCoverageReportToAudits(
      coverageWithMultipleIssues,
      {},
    );
    expect(result).toMatchSnapshot();
  });

  it('should prioritize onlyAudits over skipAudits when both are provided', () => {
    const result = trasformCoverageReportToAudits(mockCoverageResult, {
      onlyAudits: ['functions-coverage'],
      skipAudits: ['functions-coverage'],
    });
    expect(result).toMatchSnapshot();
  });
});
