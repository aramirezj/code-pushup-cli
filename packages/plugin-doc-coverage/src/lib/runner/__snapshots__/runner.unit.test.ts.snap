// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trasformCoverageReportToAudits > should filter audits when onlyAudits is provided 1`] = `
[
  {
    "details": {
      "issues": [
        {
          "message": "Missing documentation",
          "severity": "warning",
          "source": {
            "file": "test.ts",
            "position": {
              "startLine": 10,
            },
          },
        },
      ],
    },
    "displayValue": "75 %",
    "score": 0.75,
    "slug": "functions-coverage",
    "value": 75,
  },
]
`;

exports[`trasformCoverageReportToAudits > should filter audits when skipAudits is provided 1`] = `
[
  {
    "details": {
      "issues": [
        {
          "message": "Missing documentation",
          "severity": "warning",
          "source": {
            "file": "test.ts",
            "position": {
              "startLine": 10,
            },
          },
        },
      ],
    },
    "displayValue": "75 %",
    "score": 0.75,
    "slug": "functions-coverage",
    "value": 75,
  },
]
`;

exports[`trasformCoverageReportToAudits > should handle coverage result with multiple issues 1`] = `
[
  {
    "details": {
      "issues": [
        {
          "message": "Missing documentation",
          "severity": "warning",
          "source": {
            "file": "test1.ts",
            "position": {
              "startLine": 10,
            },
          },
        },
        {
          "message": "Missing documentation",
          "severity": "warning",
          "source": {
            "file": "test2.ts",
            "position": {
              "startLine": 20,
            },
          },
        },
      ],
    },
    "displayValue": "50 %",
    "score": 0.5,
    "slug": "functions-coverage",
    "value": 50,
  },
]
`;

exports[`trasformCoverageReportToAudits > should handle empty coverage result 1`] = `[]`;

exports[`trasformCoverageReportToAudits > should prioritize onlyAudits over skipAudits when both are provided 1`] = `
[
  {
    "details": {
      "issues": [
        {
          "message": "Missing documentation",
          "severity": "warning",
          "source": {
            "file": "test.ts",
            "position": {
              "startLine": 10,
            },
          },
        },
      ],
    },
    "displayValue": "75 %",
    "score": 0.75,
    "slug": "functions-coverage",
    "value": 75,
  },
]
`;

exports[`trasformCoverageReportToAudits > should transform coverage report to audit outputs with no filters 1`] = `
[
  {
    "details": {
      "issues": [
        {
          "message": "Missing documentation",
          "severity": "warning",
          "source": {
            "file": "test.ts",
            "position": {
              "startLine": 10,
            },
          },
        },
      ],
    },
    "displayValue": "75 %",
    "score": 0.75,
    "slug": "functions-coverage",
    "value": 75,
  },
  {
    "details": {
      "issues": [],
    },
    "displayValue": "100 %",
    "score": 1,
    "slug": "classes-coverage",
    "value": 100,
  },
]
`;
